src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -23,34 +25,49 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
                     return element!;
                 }
             }
-            else
+            else if (source is IList<TSource> list)
```

Comment:
Is there any reason why we couldn't incorporate the `IPartition<T>` and `IList<T>` checks in the `TryGetElement` method? We won't surface the interface exception for out-of-bounds IList lookup but perhaps that's preferable.

Reply:
Because the detection has been done before calling `TryGetElement`. Anyway, I updated the code to consolidate everything into one `TryGetElementAt` method.

Please notice the the existing `ElementAt(int)`/`ElementAtOrDefault(int)` becomes slightly different when index is out of bounds:
- Before this change, `ElementAt(int)` does not detect the count in any way, if index >= count, it enumerates all elements and throws exception. 
  - After this change, it detects count, if index >= count, it does not enumerate any element.
- Before this change, `ElementAtOrDefault(int)` only detects the count for `IList<T>`. If index >= count and the source is `ICollection<T>`/`ICollection`/`IListProvider<T>`, it enumerates all elements and return false. 
  - After this change, it detects count from all those types, if index >= count, it does not enumerate any element.
- Before this change, If the source is `IList<T>` and index is out of bounds, `ElementAt` directly pass index to list[index], let the list's indexer throw exception
  - After this change, if the source is `IList<T>` and index is out of bounds, `ElementAt` directly throws exception, does not pass index to the list[index].

Reply:
The changes sound reasonable to me. Not enumerating certain types could technically be considered a breaking change when side-effects are involved (`Select*Iterator` types jump to mind) and we have reverted optimizations for that reason in the past. I do think however that `ElementAt` calls that induce side-effects are asking for trouble regardless.

Copying @stephentoub for an opinion on this.

Reply:
Those changes caused unit test failures of other libraries. I will cancel those changes and keep the original behavior. Please see my code update.

Reply:
What types of breaks? Were they related to not surfacing the `IList` indexer exception or not enumerating `IListProvider<T>` types?

Reply:
The breaks are all related to the 3 changes I mentioned above. No problem found related to `IListProvider<T>`.

Reply:
Now `ElementAt(Index)` have the same behavior as `ElementAt(int)`, except one case. Assume `source` is a non-countable enumerable, and `N` happens to be the count of the `source`:
- `source.ElementAt(N)` will enumerate all elements and throw an exception.
- `source.ElementAt(N)` is equivalent to `source.ElementAt(^0)`. If user calls `source.ElementAt(^0)` instead, it will not enumerate any element and directly throw an exception.

@eiriktsarpalis @stephentoub Will this be a problem?

Reply:
I think it's reasonable, `source.ElementAt(^0)` is much like calling `source.ElementAt(-1)`, so throwing immediately is acceptable behaviour.

Reply:
> The breaks are all related to the 3 changes I mentioned above. No problem found related to IListProvider<T>.

Based on the tests that you added, am I right to assume the breaks were only related to not surfacing the IList indexer exceptions and _not_ new sources not being enumerated?

Reply:
Correct.

We do not have any unit tests to verify whether the source is being enumerated by ElementAt, until I added them recently.